contracts/StakeManagerV2.sol

```diff
@@ -0,0 +1,343 @@
+        returns (uint256)
+    {
+        uint256 amountToMint = convertBnbToBnbX(msg.value);
+        _delegate();
+        BNBX.mint(msg.sender, amountToMint);
+
+        emit DelegateReferral(msg.sender, msg.value, amountToMint, _referralId);
+        return amountToMint;
+    }
+
+    /// @notice Request to withdraw BnbX and get BNB back.
+    /// @param _amount The amount of BnbX to withdraw.
+    /// @return The index of the withdrawal request.
+    function requestWithdraw(uint256 _amount) external override whenNotPaused nonReentrant returns (uint256) {
+        if (_amount == 0) revert ZeroAmount();
+
+        withdrawalRequests.push(
+            WithdrawalRequest({ user: msg.sender, amountInBnbX: _amount, claimed: false, batchId: 0, processed: false })
+        );
+
+        uint256 requestId = withdrawalRequests.length - 1;
+        userRequests[msg.sender].push(requestId);
+
+        BNBX.transferFrom(msg.sender, address(this), _amount);
+        emit RequestedWithdrawal(msg.sender, _amount);
+
+        return requestId;
+    }
+
+    /// @notice Start the batch undelegation process.
+    /// @param _batchSize The size of the batch.
+    /// @param _operator The address of the operator to undelegate from.
+    function startBatchUndelegation(
+        uint256 _batchSize,
+        address _operator
+    )
+        external
+        override
+        whenNotPaused
+        nonReentrant
+        onlyRole(OPERATOR_ROLE)
+    {
+        if (_operator == address(0)) {
+            // if operator is not provided, use preferred operator
+            _operator = OPERATOR_REGISTRY.preferredWithdrawalOperator();
+        }
+        if (!OPERATOR_REGISTRY.operatorExists(_operator)) {
+            revert OperatorNotExisted();
+        }
+
+        address creditContract = STAKE_HUB.getValidatorCreditContract(_operator);
+        uint256 pooledBnb = IStakeCredit(creditContract).getPooledBNB(address(this));
+
+        uint256 cummulativeBnbXToBurn;
+        uint256 processedCount;
+        uint256 amountInBnbXToBurn;
+        while (firstUnprocessedUserIndex < withdrawalRequests.length && processedCount < _batchSize) {
+            cummulativeBnbXToBurn = amountInBnbXToBurn + withdrawalRequests[firstUnprocessedUserIndex].amountInBnbX;
+            if (pooledBnb >= convertBnbXToBnb(cummulativeBnbXToBurn)) {
+                amountInBnbXToBurn = cummulativeBnbXToBurn;
+                withdrawalRequests[firstUnprocessedUserIndex].processed = true;
+                withdrawalRequests[firstUnprocessedUserIndex].batchId = batchWithdrawalRequests.length;
+                processedCount++;
+                firstUnprocessedUserIndex++;
+            } else {
+                break;
+            }
+        }
+
+        if (amountInBnbXToBurn == 0) revert NoWithdrawalRequests();
+        uint256 amountToWithdrawFromOperator = convertBnbXToBnb(amountInBnbXToBurn);
+        totalDelegated -= amountToWithdrawFromOperator;
+
+        batchWithdrawalRequests.push(
```

Comment:
Can we make this into a map with withdrawalBatchId as the key?
Otherwise, `completeBatchUndelegation` will keep piling on batch requests.

Reply:
Or provide an estimate on in what order this will become a problem.

Reply:
It will only become a problem if someone want to fetch the whole array in one go. 
Since we are accessing it through index and limited size, it won't be a problem.
And for the problem case above, it can be solved via pagination call.

Reply:
I also agree, this is not gonna be problem as 
- ethereum has big storage limit of 2^256 
- and above that we have maxActiveUserRequest limits, which will also help mitigate it.
```
Also with current design it is not possible to switch to mapping, 
as that would mean to delete the batch entry once it is claimed.
But then users won't be able to claim as we need the claimed batch data as well.
```